plugins/nf-prov/src/main/nextflow/prov/WrrocRenderer.groovy

```diff
@@ -0,0 +1,960 @@
+import java.nio.file.StandardCopyOption
+import java.nio.file.attribute.BasicFileAttributes
+import java.time.LocalDateTime
+import java.time.format.DateTimeFormatter
+
+import groovy.json.JsonOutput
+import groovy.transform.CompileStatic
+import nextflow.Session
+import nextflow.config.ConfigMap
+import nextflow.file.FileHolder
+import nextflow.processor.TaskProcessor
+import nextflow.processor.TaskRun
+import nextflow.script.params.FileInParam
+import nextflow.script.params.FileOutParam
+import nextflow.script.ScriptMeta
+import org.apache.commons.io.FilenameUtils
+import org.yaml.snakeyaml.Yaml
+
+/**
+ * Renderer for the Provenance Run RO Crate format.
+ *
+ * @author Ben Sherman <bentshermann@gmail.com>
+ * @author Felix Bartusch <felix.bartusch@uni-tuebingen.de>
+ * @author Famke Bäuerle <famke.baeuerle@qbic.uni-tuebingen.de>
+ */
+@CompileStatic
+class WrrocRenderer implements Renderer {
+
+    private Path path
+
+    private boolean overwrite
+
+    @Delegate
+    private PathNormalizer normalizer
+
+    // The final RO-Crate directory
+    private Path crateRootDir
+    // Nextflow work directory
+    private Path workdir
+    // Nextflow pipeline directory (contains main.nf, assets, etc.)
+    private Path projectDir
+    private Map agent
+    private Map organization
+    // List of contactPoints (people, organizations) to be added to ro-crate-metadata.json
+    private List<Map> contactPoints = []
+    private String publisherID
+
+    WrrocRenderer(Map opts) {
+        path = opts.file as Path
+        overwrite = opts.overwrite as Boolean
+
+        ProvHelper.checkFileOverwrite(path, overwrite)
+    }
+
+    private static final List<String> README_FILENAMES = List.of("README.md", "README.txt", "readme.md", "readme.txt", "Readme.md", "Readme.txt", "README")
+
+    @Override
+    void render(Session session, Set<TaskRun> tasks, Map<Path,Path> workflowOutputs) {
+        final params = session.params
+        final configMap = session.config
+
+        // Set RO-Crate Root and workdir
+        this.crateRootDir = Path.of(params['outdir'].toString()).toAbsolutePath()
+        this.workdir = session.getWorkDir()
+        this.projectDir = session.getWorkflowMetadata().getProjectDir()
+
+        // get workflow inputs
+        final taskLookup = ProvHelper.getTaskLookup(tasks)
+        final workflowInputs = ProvHelper.getWorkflowInputs(tasks, taskLookup)
+
+        // Add intermediate input files (produced by workflow tasks and consumed by other tasks)
+        workflowInputs.addAll(getIntermediateInputFiles(tasks, workflowInputs))
+        final workflowInputMapping = getWorkflowInputMapping(workflowInputs)
+
+        // Add intermediate output files (produced by workflow tasks and consumed by other tasks)
+        workflowOutputs.putAll(getIntermediateOutputFiles(tasks, workflowOutputs))
+
+        // Copy workflow input files into RO-Crate
+        workflowInputMapping.each { source, dest ->
```

Comment:
It looks like you are copying the intermediate files into the RO-crate? I don't think this is feasible in general. I think it would be better to save only a record of the task inputs/outputs with a checksum. That is what the BCO does at least.

Reply:
Recording `object` (inputs) and `result` (outputs) for `CreateAction` is respectively a MAY and a SHOULD in WRROC, so it's not strictly required. Also, workflow inputs and outputs are more important than intermediate files. However, if intermediate inputs/outputs are available, it is appropriate to include them in the RO-Crate, so that it provides a more complete representation of the run. If this is not feasible in general, could it be controlled by a boolean option (e.g. `include_intermediates`) in the configuration file?

Reply:
I have kept the `object` and `result` metadata without copying the actual intermediate files. This way you can still e.g. construct the provenance graph. I will consider adding a flag to copy the intermediates in a future iteration